Move data structures into SpeziFoundation

Supereg opened the pull request:
# Move data structures into SpeziFoundation

## :recycle: Current situation & Problem
This PR moves data structures into SpeziFoundation. See https://github.com/StanfordSpezi/SpeziFoundation/pull/2.


## :gear: Release Notes 
* Move `SharedRepository` data structure into [SpeziFoundation](https://github.com/StanfordSpezi/SpeziFoundation).


## :books: Documentation
* Documentation was reduced.


## :white_check_mark: Testing
--


## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

Supereg commented:
Looking to still include the documentation refactoring we discussed in this PR.

PSchmiedmayer commented:
@Supereg Thank you; that would be great!

PSchmiedmayer reviewed (approved):
Nice cleanup and improvement of Spezi @Supereg!

Supereg commented:
@PSchmiedmayer I have rewritten the documentation of `Module` for more clarity and to be much more structured.
* I introduced three new articles in the `Capabilities` section of `Module`: `Interactions with SwiftUI`, `Module Dependency`, `Module Communication`. Most notably:
* I have rewritten some parts, removed or added certain sections to improve clarity of documentation.
* I made sure that every property wrapper that is usable in `Module` consistently provides a note (or important note) providing clear instructions what restrictions apply (e.g., at which point the wrapped value is accessible).

Would love to hear your feedback about these changes If you have a free minute.

PSchmiedmayer commented on Sources/Spezi/Spezi.docc/Module/Interactions with SwiftUI.md:
```diff
@@ -0,0 +1,106 @@
+}
+
+
+class ContentView: Module { // access the model within SwiftUI
+    @Environment(ExampleModel.self) var model
+}
```
I would suggest to maybe split up this code example into smaller blocks with a smaller text between them instead of using the comments here. And Module -> View.
```suggestion
class ContentView: View { // access the model within SwiftUI
    @Environment(ExampleModel.self) var model
}
```

Supereg replied:
good catch 🚀

PSchmiedmayer commented on Sources/Spezi/Spezi.docc/Module/Interactions with SwiftUI.md:
```diff
@@ -0,0 +1,106 @@
+* Providing access to a global model state. For more information, have a look at [Share model data throughout a view hierarchy](https://developer.apple.com/documentation/swiftui/managing-model-data-in-your-app#Share-model-data-throughout-a-view-hierarchy).
+* Display UI components using modifiers like [alert(_:isPresented:presenting:actions:message:)](https://developer.apple.com/documentation/swiftui/view/alert(_:ispresented:presenting:actions:message:)-8584l)
+
+> Note: We strongly advise using the new `@Observable` macro instead of the previous `ObservableObject` protocol to achieve optimal performance and
```
Tip might be more applicable here.

Might also apply to some other boxes. Kind of like the more colorful and positive approach of a Tip.
```suggestion
> Tip: We strongly advise using the new `@Observable` macro instead of the previous `ObservableObject` protocol to achieve optimal performance and
```

PSchmiedmayer commented on Sources/Spezi/Spezi.docc/Module/Interactions with SwiftUI.md:
```diff
@@ -0,0 +1,106 @@
+
+By adopting the ``LifecycleHandler``, your `Module` can provide lifecycle methods to the underlying `UIApplication` and ``Spezi/Spezi`` lifecycle.
+
+### Module Constraints
```
I think it might make sense for this section to move the main module documentation?

PSchmiedmayer commented on Sources/Spezi/Spezi.docc/Module/Module Communication.md:
```diff
@@ -0,0 +1,55 @@
+Before the invocation of ``Module/configure()-5pa83``, the data of all ``Module/Provide`` properties will be made available.
+Refer to the documentation of the property wrappers for a more detailed overview of the available capabilities.
+
+> Important: Accessing `@Provide` properties within the ``Module/configure()-5pa83`` method or accessing `@Collect` properties before
+``Module/configure()-5pa83``, will result in a runtime error. 
```
It might make sense to formulate this a bit more positive and an instruction how to do this rather then an example how not to do it.

Maybe like: Hey, you can access these things in the collect and these after collect. Anything else will result in a runtime crash.

Random question, why can I not access the properties in the collect method? I should write them in the initializer but reading them should work in configure shouldn't it. Or am I missing something there?

PSchmiedmayer commented on Sources/Spezi/Spezi.docc/Module/Module Communication.md:
```diff
@@ -0,0 +1,55 @@
+class ModuleC: Module {
+    @Collect var allGreetings: [String]
+
+    func collect() {
+        print("All the greetings we received: \(allGreetings)") // prints "Hello", "Hola", "Hallo" in any order
+    }
```
Might be cool to use the configure method here to demonstrate the usage?

Supereg replied:
Actually was the intention here. Probably confused something there 😅

PSchmiedmayer commented on Sources/Spezi/Spezi.docc/Module/Module Communication.md:
```diff
@@ -0,0 +1,55 @@
+    @Provide var someGreeting = "Hello"
+}
+
+class ModuleB: Module {
+    @Provide var someGreeting = ["Hola", "Hallo"]
+}
```
Might be cool to see the initializer being used to set the provide elements to provide an other example?

PSchmiedmayer commented on Sources/Spezi/Spezi.docc/Module/Module Dependency.md:
```diff
@@ -0,0 +1,53 @@
+If you declare a dependency to a `Module` that is not configured by the users (e.g., some underlying configuration `Module`s might not event be
+publicly accessible), is initialized with the instance that was passed to the ``Module/Dependency`` property wrapper.
+
+- Note: `Module`s can easily provide a default configuration by adopting the ``DefaultInitializable`` protocol.  
```
Might be cool to make this a Tip and maybe provide an example here.

PSchmiedmayer commented on Sources/Spezi/Standard/Module+Standard.swift:
```diff
@@ -16,16 +16,17 @@ extension Module {
+    ///
+    /// A ``Module`` can define the injection of a ``Standard`` using the @``Module/StandardActor`` property wrapper.
+    ///
+    /// - Note: You can access `@StandardActor` once your ``Module/configure()-5pa83`` method is called (e.g., it must not be used in the `init`)
```
Might be cool to provide this as a tip in the main Module documentation, I think this could be a common use case.

PSchmiedmayer reviewed (approved):
Thank you for all the improvements. I had some general comments on how to improve the documentation; it is already great to see the new structure and how this does a good job demonstrating the usage of the Module type and the property wrappers we have there!

I am happy to see the PR merged once the conversations are resolved.
